Source/WebCore/loader/DocumentPrefetcher.h

```diff
@@ -0,0 +1,75 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "CachedResourceClient.h"
+#include "CachedResourceHandle.h"
+#include <wtf/CompletionHandler.h>
+#include <wtf/Forward.h>
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+#include <wtf/URLHash.h>
+
+
+namespace WebCore {
+
+class CachedRawResource;
+class DocumentLoader;
+class FrameLoader;
+class ResourceRequest;
+class NetworkLoadMetrics;
+
+class DocumentPrefetcher : public RefCounted<DocumentPrefetcher>, public CachedRawResourceClient {
+public:
+    static Ref<DocumentPrefetcher> create(FrameLoader& frameLoader) { return adoptRef(*new DocumentPrefetcher(frameLoader)); }
+    explicit DocumentPrefetcher(FrameLoader&);
+    ~DocumentPrefetcher();
+
+    void prefetch(const URL&, const Vector<String>& tags, const String& referrerPolicyString, bool lowPriority = false);
+
+    // CachedRawResourceClient
+    void responseReceived(const CachedResource&, const ResourceResponse&, CompletionHandler<void()>&&) override;
+    void redirectReceived(CachedResource&, ResourceRequest&&, const ResourceResponse&, CompletionHandler<void(ResourceRequest&&)>&&) override;
+    void notifyFinished(CachedResource&, const NetworkLoadMetrics&, LoadWillContinueInAnotherProcess = LoadWillContinueInAnotherProcess::No) override;
+    CachedResourceClientType resourceClientType() const override { return RawResourceType; }
+
+    bool isFinished(const URL& url) const { return m_finishedURLs.contains(url); }
+
+    void notifyWhenFinished(const URL& url) { m_notifyWhenFinishedURLs.add(url); }
+    bool isNotifyingWhenFinished(const URL& url) const { return m_notifyWhenFinishedURLs.contains(url); }
+
+private:
+    void clear();
+    void clearPrefetchedAssets();
+
+    WeakRef<FrameLoader> m_frameLoader;
+    HashMap<URL, CachedResourceHandle<CachedRawResource>> m_prefetchedResources;
+    HashMap<URL, Box<NetworkLoadMetrics>> m_prefetchedNetworkLoadMetrics;
```

Comment:
Can this be 1 map of a URL to a structure that contains a CachedResourceHandle and a NetworkLoadMetrics and a bool indicating whether we should notify when finished?